samples/bpf/io_filter_user.c

```diff
@@ -10,24 +10,43 @@
 #include "trace_helpers.h"
 #include "bpf_load.h"
 
+//pass in device as argument (ie. /dev/sda)
+
 int main(int argc, char **argv)
 {
+	struct bpf_object *obj;
 	char filename[256], test_file[256];
-	int fd, ret;
+	int fd, ret, progfd, devfd;
 
 	snprintf(filename, sizeof(filename), "%s_kern.o", argv[0]);
 
-	if (load_bpf_file(filename)) {
-		printf("error loading program: %s\n", bpf_log_buf);
+	ret = bpf_prog_load(filename,
+			    BPF_PROG_TYPE_IO_FILTER, &obj, &progfd);
+
+	printf("loaded submit_bio program successfully\n");
+
+	devfd = open(argv[1], O_RDWR | O_NONBLOCK);
+
+	if (devfd == -1) {
+		printf("failed to open lirc device %s: %m\n", argv[1]);
 		return 1;
 	}
 
 	printf("loaded submit_bio program successfully\n");
 
+	ret = bpf_prog_attach(progfd, devfd, BPF_BIO_SUBMIT, 0);
+
+	if (ret) {
+		printf("Failed to attach bpf submit_bio program to device: %m\n");
+		return 1;
+	}
+
+	printf("attached bpf submit_bio program to device\n");
+
 	strcpy(test_file, getenv("HOME"));
 	strcat(test_file, "/submit_bio_test");
 
-	fd = open(test_file, O_WRONLY|O_CREAT|O_TRUNC, S_IRWXU | S_IRWXG | S_IRWXO);
+	fd = open(test_file, O_WRONLY|O_CREAT|O_TRUNC, S_IRWXU|S_IRWXG|S_IRWXO);
```

Comment:
Can we drop the mode argument?

Reply:
Open manpage says the mode arguments are required when O_CREAT is used. However, it does indeed work (can write to file) when mode argument is dropped, but looking at file after running, permissions are "----------"